Unicode/ustring.c

```diff
@@ -345,72 +358,79 @@ return( u_copy( s1 ));
 }
 
-unichar_t *uc_copyn(const char *pt,int len) {
-    unichar_t *res, *rpt;
-
-    if(!pt)
-return((unichar_t *)0);
+unichar_t *uc_copyn(const char *pt, int n) {
+/* Copy n unsigned chars{0..255} to a unichar_t type */
+/* string. If not printable, return an empty string. */
+    unichar_t *res, *rpt, t;
 
+    if ( pt && ++n>0 ) {
 #ifdef MEMORY_MASK
-    if ( (len+1)*sizeof(unichar_t)>=MEMORY_MASK )
-	len = MEMORY_MASK/sizeof(unichar_t)-1;
+	if ( n>MEMORY_MASK/sizeof(unichar_t) )
+	    n = MEMORY_MASK/sizeof(unichar_t);
 #endif
-    res = (unichar_t *) malloc((len+1)*sizeof(unichar_t));
-    for ( rpt=res; --len>=0 ; *rpt++ = *(unsigned char *) pt++ );
-    *rpt = '\0';
-return(res);
+	if ( (res=rpt=(unichar_t *)(malloc(n*sizeof(unichar_t)))) ) {
+	    if ( (--n) ) {
+		while ( --n>=0 ) {
+		    t = (unichar_t)(*(unsigned char *) pt++);
+		    *rpt++ = t & 0xff;
+		}
+	    }
+	    *rpt = 0;
+	    return( res );
+	}
+    }
+    return( calloc(1,sizeof(unichar_t)) );
 }
 
 unichar_t *uc_copy(const char *pt) {
-    unichar_t *res, *rpt;
+/* Copy unsigned char{0..255} string to unichar type */
+/* string. If not printable, return an empty string. */
     int n;
 
-    if(!pt)
-return((unichar_t *)0);
+    if ( pt && (n=strlen(pt))>0 )
+	return( uc_copyn(pt,n) );
 
-    n = strlen(pt);
-#ifdef MEMORY_MASK
-    if ( (n+1)*sizeof(unichar_t)>=MEMORY_MASK )
-	n = MEMORY_MASK/sizeof(unichar_t)-1;
-#endif
-    res = (unichar_t *) malloc((n+1)*sizeof(unichar_t));
-    for ( rpt=res; --n>=0 ; *rpt++ = *(unsigned char *) pt++ );
-    *rpt = '\0';
-return(res);
+    return( calloc(1,sizeof(unichar_t)) );
 }
 
-char *cu_copyn(const unichar_t *pt,int len) {
+char *cu_copyn(const unichar_t *pt, int n) {
+/* Copy unichar string to str. Assume values 0..255. */
     char *res, *rpt;
+    unichar_t t;
 
-    if(!pt)
-return(NULL);
-
+    if ( pt && ++n>0 ) {
 #ifdef MEMORY_MASK
-    if ( (len+1)>=MEMORY_MASK )
-	len = MEMORY_MASK-1;
+	if ( n>MEMORY_MASK )
+	    n = MEMORY_MASK;
 #endif
-    res = (char *) malloc(len+1);
-    for ( rpt=res; --len>=0 ; *rpt++ = *pt++ );
-    *rpt = '\0';
-return(res);
+	if ( (res=rpt=malloc(n)) ) {
+	    if ( (--n) ) {
+		while ( --n>=0 ) {
+		    if ( (t=*pt++)>0xff ) {
+			free(res);
+			return( NULL );
+		    }
+		    *rpt++ = t;
+		}
+	    }
+	    *rpt = '\0';
+	    return( res );
+	}
+    }
+    return( NULL );
 }
 
 char *cu_copy(const unichar_t *pt) {
-    char *res, *rpt;
+/* Copy unichar to char string, assume values 0..255 */
+/* and if error, return( "" ) which can be made free */
+    char *res;
     int n;
 
-    if(!pt)
-return((char *)0);
+    if ( pt && (n=u_strlen(pt))>0 )
+	if ( (res=cu_copyn(pt,n))!=NULL )
+	    return( res );
 
-    n = u_strlen(pt);
-#ifdef MEMORY_MASK
-    if ( (n+1)>=MEMORY_MASK )
-	n = MEMORY_MASK/sizeof(unichar_t)-1;
-#endif
-    res = (char *) malloc(n+1);
-    for ( rpt=res; --n>=0 ; *rpt++ = *pt++ );
-    *rpt = '\0';
-return(res);
+    return( calloc(1,sizeof(unichar_t)) );
```

Comment:
As elsewhere, we need to be sure that this effective change in behavior (returning dynamically allocated `""` instead of `NULL` when given a NULL pointer) doesn't break anything. @JoesCat knows more about the use cases than I do.

Reply:
I mentioned return( "" ), but we're talking about calloc(1,sizeof(unichar_t)).
The code of interest is return( (unichar_t)0 )